### ISSUE TITLE
Implement JPA 2.0 Criteria Queries

### ISSUE THREAD

**Issue by masha256** [Impetus/kundera#118]
I find it misleading that kundera claims "Kundera is a JPA 2.0 compliant object-datastore mapping library for NoSQL datastores. " when it doesn't support criteria queries. Criteria queries are a formal part of the JPA 2.0 specification. Not supporting them means you are not JPA 2.0 compliant.

**Comment by mevivs** [comment]
Yes it is JPA compliant but not 100%. Purpose to build kundera is not just another ORM solution, but to leverage it for NOSQL.
However work on CriteriaQuery support is planned.

-Vivek

**Comment by xamry** [comment]
You are right that Kundera isn't "fully JPA compliant" in that sense. Maybe we could put a "*" and write "partial compliant" as a footnote :-)
In fact writing something that follows a specification 100% could be an overkill for NoSQL IMHO. Since Kundera is more inclined towards NoSQL, stress here is to provide JPA way for almost anything that is required while working with NoSQL. So many things like criteria query are desirable, but have another alternative available, so they are a less priority at this moment. This is an ongoing process. Kundera maybe considered at an intermediate stage in that pursuit. ORMs like Hibernate were built to solve 80% of the most frequently used Object database impedance mismatch problem. For a full low level database functionality, people preferred plain JDBC.

**Comment by masha256** [comment]
I am working on a project (that will hopefully be open source soon) that allows enterprises to use one API for both RDBMs and NoSQL (cassandra at the moment). I am implementing the query system and would  like to standardize on JPA 2.0 criteria queries as that API and was looking at Kundera as the JPA bridge to the NoSQL storage engines. Without criteria queries, I am not sure if kundera will do what we want. I am going to spend time researching perhaps putting QueryDSL in front of Kundera since QueryDSL has its own query API that compiles down to JPQL and not criteria queries.

**Comment by mevivs** [comment]
I understand your point. Just to clarify that, Kundera supports JPQL for many clauses, but may not work for things like JOIN clause etc. So hooking up a transformation should work for you. I will post an update, once we implement CriteriaQuery interface, as mentioned earlier it is already planned.

Thanks,
Vivek

**Comment by pires** [comment]
Criteria is awesome when you have a static-metamodel (strong-typed queries with compile-time evaluation) and I think Kundera doesn't support this as well. Would be great to have this!

**Comment by poornerd** [comment]
I would love to see this, too. Is there an ETA?

**Comment by mevivs** [comment]
First cut support added for JPA criteria 2.0 but still under development. Though Basic JPA Criteria query support should work.

-Vivek

**Comment by prasad14** [comment]
Does kundera 2.14 support criteria query?

**Comment by chhavigangwal** [comment]
@prasad14 

Basic criteria query support works with Kundera-2.14. A sample test case can be found at :         

Criteria related queries can be found in  

https://github.com/impetus-opensource/Kundera/blob/273c13342ddd1aceed0cd23504649926ce8fdb84/src/kundera-cassandra/cassandra-core/src/test/java/com/impetus/client/crud/BaseTest.java

These are invoked from following test case 

https://github.com/impetus-opensource/Kundera/blob/273c13342ddd1aceed0cd23504649926ce8fdb84/src/kundera-cassandra/cassandra-ds-driver/src/test/java/com/impetus/kundera/client/cassandra/crud/CriteriaQueryTest.java

Chhavi

**Comment by prasad14** [comment]
@chhavigangwal 

But in that code I didn't see the creation of CriteriaBuilder object?
Can you  explain that scenario?

**Comment by chhavigangwal** [comment]
Please find the updated comment.

Chhavi

**Comment by prasad14** [comment]
Does kundera 2.15 support all the operators of criteria builder

**Comment by mevivs** [comment]
@prasad14  i think it will be great if you could mention what exactly you looking with Criteria builder support? This will surly help Kundera team to answer in more accurate way.

-Vivek

**Comment by prasad14** [comment]
@chhavigangwal 

does kundera support all the operators in criteria builder

**Comment by hailyj** [comment]
Im trying to work with Kundera 2.15 using thrift client .
Can u suggest me the list of jars which i have to use?

**Comment by chhavigangwal** [comment]
@hailyj 

Since Kundera is a maven based project , you need to include the following dependency in pom of your project to start using Kundera's  Cassandra module :

```
      <dependency>
            <groupId>com.impetus.kundera.client</groupId>
            <artifactId>kundera-cassandra</artifactId>
            <version>2.15</version>
        </dependency>
```

Please use the following link for further reference : 
https://github.com/impetus-opensource/Kundera/wiki/Getting-Started-in-5-minutes

Hope that Helps!
Chhavi

**Comment by hailyj** [comment]
Though i ran with dependency jar obtained from maven repository ,Im getting the error:
No Persistence provider for EntityManager named cassandra_pu

How do I correct this?

**Comment by hailyj** [comment]
@chhavigangwal 
Though i tried with the dependency jar obtained from the maven repository,I got an error:" No Persistence provider for EntityManager named cassandra_pu"
Can you please help me with this

**Comment by chhavigangwal** [comment]
@hailyj 

Can you share a sample project or code snippet of how are you trying to create EMF object using Kundera ? 

Chhavi

**Comment by hailyj** [comment]
```
    EntityManagerFactory emf = Persistence.createEntityManagerFactory("cassandra_pu");
    EntityManager em = emf.createEntityManager();
```
